### ISSUE TITLE
Create pre-compiled executables

### ISSUE THREAD

**Issue by russel** [Me-TV/Me-TV#3]
As highlighted in Issue #2 users should not have to install the Rust compilation tool chain to run Me TV. A way of getting the continuous integration servers to deliver the executables so they can be downloaded by users should be put in place.

**Comment by tomidis** [comment]
Compiling with Rust v1.25 for me gets a lot of the following errors (if I remember correctly, exactly 19 of them):
`error[E0658]: 128-bit type is unstable (see issue #35118)`
Compiling with Rust v1.29 is successfully, but only after installing the missing dependencies (libgtk2.0-dev, libgstreamer1.0-dev, libgtk-3-dev)

**Comment by russel** [comment]
The compilation with 1.25 probably never got to the stage of worrying about missing dependencies. I believe you would need some of those packages anyway for compiling the code. You shouldn't need libgtk2.0-dev. You will then also need the gstreamer1.0 plugin packages. I have base, good, bad, and ugly installed, but also the gtk plugin which has been separated out.

**Comment by dougK44883** [comment]
pre-compiled executables are **exactly** what _end-users_ need - I understand you're not there yet.  
 Getting ready to build me-tv (don't like installing all the *-dev packages I otherwise have no use for) I discovered the in readme section [Setting up for watching DVB](https://github.com/Me-TV/Me-TV/blob/master/README.md#setting-up-for-watching-dvb) references the pacakge _dvb-scan-tables_.  At least in debian it is  _dtv-scan-tables_.
  Back to the top, you reference vers 3.0.x and 3.1.x for me-tv different dependencies for gstreamer - packages on debian only go up to 1.14.4-1 not 1.14.5 needed for 3.1.x  - from [https://gstreamer.freedesktop.org/](https://gstreamer.freedesktop.org/)   _GStreamer 1.14.4 stable bug fix release 2018-10-02_  are  you using a developer release? (well, not an actual release)

**Comment by russel** [comment]
GitHub and Travis-CI do not have a good workflow when it comes to pre-compiled executables, GitLab has a much better offering. However, we are here. But yes, end users should not have to install developer packages just to make use of a product, so pre-compiled executables has to be a thing. I think I might try and set up a Debian (and thus Ubuntu and Mint) repository on Bintray and also have simple executable downloads there as well – one of the great advantages of Rust builds is that there are far fewer dependencies that with C or C++, so a build from Debian Sid should work on a lot of Linux platforms as long as the GTK+ and GStreamer libraries are in place. I am hoping Travis can then deploy to Bintray. I no longer have access to Fedora so can't build for that, nor Arch. This is where having a proper CI/CD workflow would come in handy the platform should be able to do all the grunt work using containers. Travis-CI is not it, GitLab might be.

Thanks for the "heads up" on package name. I'll fix that.

Me TV 3.1.x is on hold for now till GStreamer 1.14.5 is released. I made plans when having to use a full git master/HEAD build of GStreamer to do the work on the MPEG-TS stuff some weeks ago, but with that complete (for GStreamer but not the Rust binding) I then reverted back to released GStreamer to avoid any problems in developing Me TV. So for the moment GStreamer 1.14.4 is the latest GStreamer, Me TV 3.0.6 the latest Me TV, with 3.0.7 likely to happen to get the extra recording executables out there. I want to solve the pre-compiled executables to do that though if at all possible. I won't have quite so much time for this over the next few days on the grounds there are some festivities happening. :-) However, I am on the job.

**Comment by russel** [comment]
I think Bintray is the place to store pre-compiled executables. It turns put Travis-CI can now deploy build products to Bintray, so the deploy stage is feasible. I'll try that for the v3.0.7 release. For now I have manually uploaded some executables to the v3.0.6 release on Bintray, and amended the README file.

@dougK44883 if I could get you to the the "guinea pig" and try out downloading from Bintray v3.0.6, I'd appreciate it.

**Comment by dougK44883** [comment]
> @dougK44883 if I could get you to the the "guinea pig" and try out downloading from Bintray v3.0.6, I'd appreciate it.

absolutely!  I have a tuner card installed, confirmed working via me-tv 1.4x.  Updated [dvb-]channels.conf to v5 had the gstreamers packages confirmed.  Stop short of rust and build-essential type packages... until after the _festivities_  I'll get too caught up and distracted

**Comment by russel** [comment]
It appears that upload to Bintray of executables created during CI testing on Travis-CI works. Travis-CI is currently running Ubuntu Xenial, so the executables should work on most any variant of Debian, Ubuntu and Mint post 2016.

The idea will be to upload updated snapshot executables for every commit, and release executables for each tagged release. It is still all a bit manual, but I am not sure how to automate things any more than they are just now.

**Comment by russel** [comment]
Executables compiled on Ubuntu Xenial by Travis-CI now to be found on Bintray at https://bintray.com/beta/#/me-tv/Downloads/Me-TV By rights they ought to work on Fedora, Arch, etc.

**Comment by dougK44883** [comment]
Just an annoying non-issue, _me-tv -V_ yields _Me TV 0.0.0_,  trying to keep track of recently updated versions while debugging this is handy.  I did see the comment in main.rs 
`     *  Thus there is no Rust/GTK+ way of handling command line arguments.`
which seems flubbed up, but I'm not a developer.   All I can do is use a shell script from the "Set me up  Using this repo with cURL" from binytray.com to download updated me-tv* files.  Again, thanks greatly for do more than anyone knows!

**Comment by russel** [comment]
The Rust (actually gtk-rs) problem with command line arguments is annoying and wrong, it is not a GTK+ problem, it has the facility, a facility that works fine with D and GtkD. This is a gtk-rs problem. I shall fix the comment, and then go rant quietly to myself in a corner somewhere. I shall also do something about this using a non-standard, i.e. not consistent with GTK+, mechanism. Actually, it is almost certainly in place with just the wrong data being issued. i.e. this is definitely a bug and should be trivially fixable.

The about dialogue give the correct value. :-)

**Comment by russel** [comment]
OK so it was as easy as expected. Truth be told: I just forget to replace the 0.0.0 with what should have been there. :-) I have pushed the change so the pre-built executable should get an update in about 10 minutes or so.

**Comment by dougK44883** [comment]
> OK so it was as easy as expected.

cook, when I read your previous post, I didn't care to stir up too much hassle about it.  Thanks greatly.

**Comment by russel** [comment]
I am going to close this issue for now on the grounds that there is a solution that should work for most Linux users. If someone comes up with a new requirement, a new issue is probably the way of dealing with it.

**Comment by dougK44883** [comment]
> The compilation with 1.25 probably never got to the stage of worrying about missing dependencies. I believe you would need some of those packages anyway for compiling the code. You shouldn't need libgtk2.0-dev. You will then also need the gstreamer1.0 plugin packages. I have base, good, bad, and ugly installed, but also the gtk plugin which has been separated out.

Looks like _gstreamer1.0-libav_ should be added to the list of [Dependencies](https://github.com/Me-TV/Me-TV/blob/master/README.md#dependencies).  (but I had bigger problems before it became an issue)

**Comment by russel** [comment]
Duly added. Essential for _me-tv-record_ as currently implemented to work.

**Comment by dougK44883** [comment]
v3.0.8 went SOL :(
[https://dl.bintray.com/me-tv/Downloads/:me-tv](https://dl.bintray.com/me-tv/Downloads/:me-tv)
returns a file containing..
`
/home/travis/build/russel/Me-TV/target/release/me-tv: /home/travis/build/russel/Me-TV/src/about.rs /home/travis/build/russel/Me-TV/src/channel_names.rs /home/travis/build/russel/Me-TV/src/control_window.rs /home/travis/build/russel/Me-TV/src/control_window_button.rs /home/travis/build/russel/Me-TV/src/dialogs.rs /home/travis/build/russel/Me-TV/src/dvb.rs /home/travis/build/russel/Me-TV/src/frontend_manager.rs /home/travis/build/russel/Me-TV/src/frontend_window.rs /home/travis/build/russel/Me-TV/src/gstreamer_engine.rs /home/travis/build/russel/Me-TV/src/input_event_codes.rs /home/travis/build/russel/Me-TV/src/main.rs /home/travis/build/russel/Me-TV/src/metvcomboboxtext.rs /home/travis/build/russel/Me-TV/src/preferences.rs /home/travis/build/russel/Me-TV/src/preferences_dialog.rs /home/travis/build/russel/Me-TV/src/remote_control.rs /home/travis/build/russel/Me-TV/src/resources/application_menu.xml /home/travis/build/russel/Me-TV/src/resources/control_window_menu.xml /home/travis/build/russel/Me-TV/src/resources/frontend_window_fullscreen_toolbar.glade.xml /home/travis/build/russel/Me-TV/src/resources/images/me-tv.png /home/travis/build/russel/Me-TV/src/resources/preferences_dialog.glade.xml /home/travis/build/russel/Me-TV/src/transmitter_dialog.rs
`
me-tv-record is similiar.

**Comment by russel** [comment]
Authority problem in the Travis-CI → Bintray pipeline. I am awaiting data from Travis-CI to try and solve this. I'll release a 3.0.9 when they do. In the interim I'll try doing a manual drop of 3.0.8.

**Comment by russel** [comment]
3.0.8 dropped but without GPG check files.

**Comment by russel** [comment]
Clearly the Travis-CI → Bintray executable deployment is not yet working so reopening this issue.

**Comment by russel** [comment]
The problem to be solved is how to get the GitHub Me-TV organisation authorised on Travis-CI to upload executables to the Me-TV organisation on Bintray.

**Comment by russel** [comment]
It seems that GitLab provides a much more integrated and easier way of handling this than using the GitHub/Travis-CI/Bintray solution. I am putting together release publishing using GitLab to replace the current system. This may lead to switching the mainline from GitHub to GitLab.

**Comment by russel** [comment]
For now, executables are manually put onto Bintray. Trying to automate this is getting seriously hard. I am closing this issue since there are executables available. Work will continue trying to automate publication but without the stress of an open issue.
